js/service/attachmentservice.js

```diff
@@ -66,4 +71,88 @@ define(function(require) {
 		return Promise.resolve($.ajax(url));
 	}
 
+	/**
+	 * @param {File} file
+	 * @param {function} localAttachmentModel
+	 * @returns {Promise}
+	 */
+	function uploadLocalAttachment(file, localAttachmentModel) {
+		var fd = new FormData();
+		fd.append('attachment', file);
+
+		var progressCallback = localAttachmentModel.onProgress;
+		var url = OC.generateUrl('/apps/mail/attachments');
+
+		return new Promise(function(resolve, reject) {
+			$.ajax({
+				url: url,
+				type: 'POST',
+				xhr: function() {
+					var customXhr = $.ajaxSettings.xhr();
+					// save the xhr into the model in order to :
+					//  - distinguish upload and nextcloud file attachments
+					//  - keep the upload status for later use
+					localAttachmentModel.set('uploadRequest', customXhr);
+					// and start the request
+					if (customXhr.upload && _.isFunction(progressCallback)) {
+						customXhr.upload.addEventListener(
+							'progress',
+							progressCallback.bind(localAttachmentModel),
+							false);
+					}
+					return customXhr;
+				},
+				data: fd,
+				processData: false,
+				contentType: false
+			})
+			.done(function(data) {
+				resolve(data.id);
+			})
+			.fail(function(err) {
+				reject();
+			});;
+		});
+	}
+
+	/**
+	 * This method is called when a local attachment upload should be aborted.
+	 * If there is no upload ongoing, this method has no effect.
+	 *
+	 * @param {function} localAttachmentModel
+	 */
+	function abortLocalAttachment(localAttachmentModel) {
+		var uploadRequest = localAttachmentModel.get('uploadRequest');
+		if (uploadRequest && uploadRequest.readyState < 4) {
+			uploadRequest.abort();
```

Comment:
Would make sense to unset the `uploadRequest` property here if we don't need it anymore.